engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskCommandPreconditionChecker.java

```diff
@@ -52,13 +58,32 @@ protected Either<Tuple<RejectionType, String>, UserTaskRecord> check(
 
     final LifecycleState lifecycleState = userTaskState.getLifecycleState(userTaskKey);
 
-    if (validLifecycleStates.contains(lifecycleState)) {
-      return Either.right(persistedRecord);
+    if (!validLifecycleStates.contains(lifecycleState)) {
+      return Either.left(
+          Tuple.of(
+              RejectionType.INVALID_STATE,
+              String.format(INVALID_USER_TASK_STATE_MESSAGE, intent, userTaskKey, lifecycleState)));
+    }
+
+    if (intent.equals(CLAIM_INTENT)) {
+      final String assignee = command.getValue().getAssignee();
+      if (assignee.isBlank()) {
+        return Either.left(
+            Tuple.of(
+                RejectionType.INVALID_STATE,
+                String.format(INVALID_USER_TASK_EMPTY_ASSIGNEE_MESSAGE, intent, userTaskKey)));
+      }
+
+      final String currentAssignee = persistedRecord.getAssignee();
+      final boolean canClaim = currentAssignee.isBlank() || currentAssignee.equals(assignee);
+      if (!canClaim) {
+        return Either.left(
+            Tuple.of(
+                RejectionType.INVALID_STATE,
+                String.format(INVALID_USER_TASK_ASSIGNEE_MESSAGE, intent, userTaskKey)));
+      }
     }
```

Comment:
:wrench: Maybe we can extract all the Claim-related guard clauses into a separate `private` method that returns an `Either`. That way we can chain it to other methods that we add in the future, making the code more readable.

I'm suggesting this as we may add additional validations here, and the code will quickly get messy.

Reply:
Totally agree, will make this change today 👍🏻

Reply:
💭 A bit late to the party, but would it make sense to keep the command-specific checks in the command processors as methods and hand them over to the `check` method as reference or anything similar? We'll get more checks over time and I'm not sure if it's ideal to decide on the specific checks based on the name of the intent in the checker class. Haven't tried it in code yet but maybe that makes sense to you as well. Our maybe you've tried this and it felt clunky. Happy for any thoughts here.

Reply:
Thank you for the suggestion, I had the idea to have all the checks in the same place instead of having part of the checks in the processor and another part in `UserTaskCommandPreconditionChecker` as it looked a bit cleaner to me and it seemed like intent specific checks also belong to that class. But may be making a decision on the specific checks based on the intent name is really not the best idea 🧐

@koevskinikola what do you think?

Reply:
I like the idea of handing over command-specific checks to the `check` method. That way we still do the checking through a single class, but the command-specific code is contained in the appropriate class.

Reply:
> I like the idea of handing over command-specific checks to the `check` method. That way we still do the checking through a single class, but the command-specific code is contained in the appropriate class.

I will create a task for myself to refactor that, thank you for the feedback.

Reply:
@koevskinikola as per our conversation I will create a `refactor` commit to accommodate these changes while working on the next task.

@tmetzke thanks for the suggestion!

Reply:
@ana-vinogradova-camunda, I'll start working on the `update` command now. I assume I can take over refactoring this and you haven't started anything in that direction, right?

Reply:
That's correct, I am currently working on the bug, so haven't started anything in that direction yet. Thank you @tmetzke

Reply:
Thanks, @ana-vinogradova-camunda, for confirming 👍 I have added the refactoring as a first step now in https://github.com/camunda/zeebe/pull/15819.